nexus/src/app/sagas/instance_create.rs

```diff
@@ -1014,3 +1014,234 @@ async fn sic_instance_ensure(
+        use crate::db::schema::instance::dsl;
+
+        dsl::instance
+            .filter(dsl::time_deleted.is_null())
+            .select(Instance::as_select())
+            .first_async::<Instance>(
+                datastore.pool_authorized(opctx).await.unwrap(),
+            )
+            .await
+            .optional()
+            .unwrap()
+            .is_none()
+    }
+
+    async fn no_network_interface_records_exist(
+        datastore: &DataStore,
+        opctx: &OpContext,
+    ) -> bool {
+        use crate::db::model::NetworkInterface;
+        use crate::db::schema::network_interface::dsl;
+
+        dsl::network_interface
+            .filter(dsl::time_deleted.is_null())
+            .select(NetworkInterface::as_select())
+            .first_async::<NetworkInterface>(
+                datastore.pool_authorized(opctx).await.unwrap(),
+            )
+            .await
+            .optional()
+            .unwrap()
+            .is_none()
+    }
+
+    async fn no_external_ip_records_exist(
+        datastore: &DataStore,
+        opctx: &OpContext,
+    ) -> bool {
+        use crate::db::model::ExternalIp;
+        use crate::db::schema::external_ip::dsl;
+
+        dsl::external_ip
+            .filter(dsl::time_deleted.is_null())
+            .select(ExternalIp::as_select())
+            .first_async::<ExternalIp>(
+                datastore.pool_authorized(opctx).await.unwrap(),
+            )
+            .await
+            .optional()
+            .unwrap()
+            .is_none()
+    }
+
+    async fn disk_is_detached(
+        datastore: &DataStore,
+        opctx: &OpContext,
+    ) -> bool {
+        use crate::db::model::Disk;
+        use crate::db::schema::disk::dsl;
+
+        dsl::disk
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::name.eq(DISK_NAME))
+            .select(Disk::as_select())
+            .first_async::<Disk>(
+                datastore.pool_authorized(opctx).await.unwrap(),
+            )
+            .await
+            .unwrap()
+            .runtime_state
+            .disk_state
+            == "detached"
+    }
+
+    async fn no_instances_or_disks_on_sled(sled_agent: &SledAgent) -> bool {
+        sled_agent.instance_count().await == 0
+            && sled_agent.disk_count().await == 0
+    }
+
+    #[nexus_test]
+    async fn test_action_failure_can_unwind(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        DiskTest::new(cptestctx).await;
+        let log = &cptestctx.logctx.log;
+
+        let client = &cptestctx.external_client;
+        let nexus = &cptestctx.server.apictx.nexus;
+        let project_id = create_org_project_and_disk(&client).await;
+
+        // Build the saga DAG with the provided test parameters
+        let opctx = cptestctx.test_opctx();
+
+        let params = new_test_params(&opctx, project_id);
+        let dag = create_saga_dag::<SagaInstanceCreate>(params).unwrap();
+
+        for i in 0..dag.get_node_count() {
```

Comment:
This is the "more automated" version of what I tried doing in https://github.com/oxidecomputer/omicron/pull/1835 - I can definitely consolidate the testing here, and create a "saga unwind tester", if that would be useful.

I basically see three phases:
1. Setup the saga. Create any orgs / projects / disks / etc needed to run the saga.
2. Iterate over every single node in the DAG, injecting failures at each one. Log which ones we're working on, to make it clear what went wrong if an error occurs. Run the saga.
3. Assert that no detritus remains, and that the system is in the same state before / after saga execution.